presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -969,7 +969,7 @@ private Partitioning selectUnionPartitioning(UnionNode node, PreferredProperties
                         .global(PreferredProperties.Global.distributed(childPartitioning.withNullsAndAnyReplicated(nullsAndAnyReplicated)))
                         .build();
                 PlanWithProperties child = node.getSources().get(sourceIndex).accept(this, childPreferred);
-                if (child.getProperties().isNodePartitionedOn(childPartitioning.getPartitioningColumns(), nullsAndAnyReplicated)) {
+                if (child.getProperties().isNodePartitionedOn(childPartitioning.getPartitioningColumns(), nullsAndAnyReplicated) && !child.getProperties().isSingleNode()) {
```

Comment:
you could add a comment here

Reply:
To me it didn't seem worth the added complexity. Also, if all children are all single distributed it doesn't mean they are on the same node.  They could be single distributed on different nodes, and not fit on one node.